File: src/main/kotlin/distributed/lock/DistributedLockAspect.java
package distributed.lock;

import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Pointcut;
import org.aspectj.lang.reflect.MethodSignature;
import org.springframework.integration.support.locks.LockRegistry;
import org.springframework.stereotype.Component;

import java.lang.reflect.Method;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.locks.Lock;


@Aspect
@Component
public class DistributedLockAspect {

    @Pointcut("@annotation(distributed.lock.DistributedLock)")
    public void distributedLock() {}


    private final LockRegistry lockRegistry;

    public DistributedLockAspect () {
        this.lockRegistry = new RedisLockRepository().getLockRegistry();
    }

    protected void acquireLock(String cacheKey, int lockTimeout) throws DistributedLockException, InterruptedException {
        Lock lock = lockRegistry.obtain(cacheKey);
        boolean isLockAcquired = lock.tryLock(lockTimeout, TimeUnit.SECONDS);
        System.out.println("Lock acquired for key: `" + cacheKey + "`:" + isLockAcquired);
        if (!isLockAcquired) {
            throw new DistributedLockException("Lock is not available for key:" + cacheKey);
        }
    }

    protected void releaseLock(String cacheKey) {
        System.out.println("Releasing lock for `" + cacheKey + "`");
        Lock lock = lockRegistry.obtain(cacheKey);
        lock.unlock();
    }

    @Around("distributedLock()")
    public Object doUnderLock(ProceedingJoinPoint pjp) throws Throwable {
        String cacheKey = getLockKey(pjp);
        int timeOut = getTimeOut(pjp);

        Object[] args = pjp.getArgs();
        Class<?> className = pjp.getTarget().getClass();
        String methodName = pjp.getSignature().getName();
        Class<?>[] argClass = ((MethodSignature) pjp.getSignature()).getParameterTypes();
        Method method = className.getMethod(methodName, argClass);
        DistributedLock annotation = method.getAnnotation(DistributedLock.class);
        String key = "" ;


        // is there a annotation : method.getParameterAnnotations()[i].length > 0
        // is any of the annotation of type :  method.getParameterAnnotations()[2][j].annotationType().name.split("\\.",0)[last element] == KeyVariable
        // pjp.getArgs[i]

        //method.getParameterAnnotations()[2][1] instanceof KeyVariable


        MethodSignature signature = (MethodSignature) pjp.getSignature();
        Method any = signature.getMethod().getAnnotation(DistributedLock.class).clazz().getMethod("getKey");
        System.out.println("----- here -----" + any.invoke(signature.getMethod().getAnnotation(DistributedLock.class).clazz().newInstance()));

        Object returnVal;

        try {
            acquireLock(cacheKey, timeOut);
            returnVal = pjp.proceed();
        } catch (DistributedLockException e) {
            e.printStackTrace();
            throw e;
        } catch (Throwable throwable) {
            // Something unexpected happened in the method (pjp)
            throwable.printStackTrace();
            throw throwable;
        } finally {
            releaseLock(cacheKey);
        }
        return returnVal;
    }

    private String getLockKey(ProceedingJoinPoint pjp) {
        MethodSignature signature = (MethodSignature) pjp.getSignature();
        return signature.getMethod().getAnnotation(DistributedLock.class).key();
    }

    private int getTimeOut(ProceedingJoinPoint pjp) {
        MethodSignature signature = (MethodSignature) pjp.getSignature();
        return signature.getMethod().getAnnotation(DistributedLock.class).timeout();
    }


}

File: src/main/kotlin/distributed/lock/DistributedLockAspect.kt
package distributed.lock

import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Around
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Pointcut
import org.aspectj.lang.reflect.MethodSignature
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.integration.support.locks.LockRegistry
import org.springframework.stereotype.Component
import java.util.concurrent.TimeUnit
import java.util.concurrent.locks.Lock

@Aspect
@Component
class DistributedLockAspect (
        @Qualifier("redis") private val lockRegistry: LockRegistry
) {

    private val log = LoggerFactory.getLogger(DistributedLockAspect::class.java)

    @Pointcut("within(@distributed.lock.DistributedLock *)")
    fun distributedLock() {
    }

    @Pointcut("execution(public * *(..))")
    fun publicMethod() {
    }

    @Around("publicMethod() && distributedLock()")
    fun doUnderLock(pjp: ProceedingJoinPoint): Any? {

        var lock : Lock? = null
        var lockAcquired = false

        try {
            val waitingTime = getWaitingTime(pjp)
            val key = getKey(pjp)

            log.info("Trying to Acquire Lock for $key")
            lock = lockRegistry.obtain(key)
            lockAcquired = lock.tryLock(waitingTime, TimeUnit.SECONDS)
            if (!lockAcquired) {
                throw Exception("Lock is not available for key:$key")
            }
            log.info("Successfully Acquired Lock")

            return runProceedingJoinPoint(pjp)

        } catch (e: DistributedProxyException) {

            log.info("Error: PJP")
            throw e.cause ?: e

        } catch (e : Exception) {

            log.info("Error: Failed to Acquire Lock ${e.message}")
            throw DistributedLockException("Failed to Acquire Lock ${e.message}")

        } finally {
            if (lockAcquired) {
                log.info("Releasing lock")
                lock?.unlock()
            }
        }
    }

    private fun runProceedingJoinPoint(pjp: ProceedingJoinPoint) : Any? {
        try {
            return pjp.proceed()
        } catch (e: Exception) {
            throw DistributedProxyException(e)
        }
    }


    private fun getKey(pjp: ProceedingJoinPoint) : String {

        val args : Array<Any> = pjp.args

        val className: Class<*> = pjp.target.javaClass

        val methodName = pjp.signature.name

        val argClass = (pjp.signature as MethodSignature).parameterTypes

        val method = className.getMethod(methodName, *argClass)

        val parameterAnnotations: Array<Array<Annotation>> = method.parameterAnnotations

        assert(args.size == parameterAnnotations.size)

        for(i in args.indices) {

            for(annotation in parameterAnnotations[i]) {

                if ((annotation is KeyVariable) && (args[i] is String) && (args[i] as String).isNotEmpty()) {

                    return args[i] as String
                }
            }
        }

        throw Exception("Invalid or Empty KeyVariable")
    }

    private fun getWaitingTime(pjp: ProceedingJoinPoint): Long {

        return pjp.target.javaClass.getAnnotation(DistributedLock::class.java).waitingTime
    }
}